Optimize unit tests

vsavkin opened the pull request:
Before: `initial 20s incremental: 5.5s` 
Before with noEmitOnError=false manually patched-in: `initial 18s incremental: 3.15s`
After: `initial 7.7s incremental: 0.5s`

It is faster for the following two reasons: 
- no es6 
- no files from node_modules go into the browser tree.

Now, the time is split more-or-less equally between Funnel, MergeTrees, and TS. 

This PR also fixed the circular dependencies check that was broken.

alexeagle commented on gulpfile.js:
```diff
@@ -302,11 +302,11 @@ gulp.task('lint', ['build.tools'], function() {
+    paths: [CONFIG.dest.js.dev.es5],
     extensions: ['.js'],
-    onParseFile: function(data) { data.src = data.src.replace(/import \* as/g, "//import * as"); }
+    onParseFile: function(data) { data.src = data.src.replace(/\/\* circular \*\//g, "//"); }
```
cheater!!

vsavkin replied:
There is no other way to do it :) The right way to do the check is to use cjs sources, not es6, and there we cannot rely on the `import as` trick.

vsavkin replied:
It is interesting that we have only one place in our code base where this is necessary. Maybe we can restructure our code to get rid of it.

alexeagle replied:
I spent a while on the other way to do it, and would love to show you :)
Will you be in the office today?

On Mon, Nov 16, 2015 at 9:20 AM Victor Savkin <notifications@github.com>
wrote:

> In gulpfile.js
> <https://github.com/angular/angular/pull/5299#discussion_r44952906>:
>
> >      extensions: ['.js'],
> > -    onParseFile: function(data) { data.src = data.src.replace(/import \* as/g, "//import * as"); }
> > +    onParseFile: function(data) { data.src = data.src.replace(/\/\* circular \*\//g, "//"); }
>
> There is no other way to do it :) The right way to do the check is to use
> cjs sources, not es6, and there we cannot rely on the import as trick.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/angular/angular/pull/5299/files#r44952906>.
>

alexeagle replied:
I worked with Tobias on that circular dep. I think we should remove it by
making it a type-checker-only dependency. I'll show you.
Here was my work on this: https://github.com/angular/angular/pull/4454

On Mon, Nov 16, 2015 at 9:21 AM Alex Eagle <alexeagle@google.com> wrote:

> I spent a while on the other way to do it, and would love to show you :)
> Will you be in the office today?
>
> On Mon, Nov 16, 2015 at 9:20 AM Victor Savkin <notifications@github.com>
> wrote:
>
>> In gulpfile.js
>> <https://github.com/angular/angular/pull/5299#discussion_r44952906>:
>>
>> >      extensions: ['.js'],
>> > -    onParseFile: function(data) { data.src = data.src.replace(/import \* as/g, "//import * as"); }
>> > +    onParseFile: function(data) { data.src = data.src.replace(/\/\* circular \*\//g, "//"); }
>>
>> There is no other way to do it :) The right way to do the check is to use
>> cjs sources, not es6, and there we cannot rely on the import as trick.
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/angular/angular/pull/5299/files#r44952906>.
>>
>

IgorMinar commented:
discussed in person. lgt

vsavkin committed:
feat(build): add an option to disable type checks when running tests

Since editors and IDEs do typechecking and show errors in place,
often there is no benefit to running type checking in our test pipeline.
This PR allows you to disable type checking:

gulp test.unit.js --noTypeChecks

This commit also makes es6 generation optional.

fix(build): removes unnecessary circular dependencies

mary-poppins commented:
Merging PR #5299 on behalf of @vsavkin to branch [presubmit-vsavkin-pr-5299](https://github.com/angular/angular/tree/presubmit-vsavkin-pr-5299).
